pkg/login/social/generic_oauth.go

```diff
@@ -209,8 +198,8 @@ func (s *SocialGenericOAuth) extractEmail(data *UserInfoJson) string {
 	}
 
 	if s.emailAttributePath != "" {
-		email := s.searchJSONForAttr(s.emailAttributePath, data.rawJSON)
-		if email != "" {
+		email, err := s.searchJSONForAttr(s.emailAttributePath, data.rawJSON)
+		if err == nil && email != "" {
```

Comment:
I think it would be best to modify extractEmail to return an error, and return the error here. If that's impractical for some reason, we should log the error.

@marefr Do you think there's any reason not to change extractEmail into returning an error?

Reply:
We could modify `extractEmail()` to return an error, but we shouldn't return error here, because there're few ways to get email. If we can't get it from `emailAttributePath`, we search it in attributes and upn.

Reply:
As discussed in chat, I think we should detect here if it's an expected error type or an unexpected one. In case it's an expected type of error (e.g. some user error), ignore it (or debug log it eventually). If it's an unexpected error however, we should at the very least log it with the error level, in case it would be bad to propagate it up the chain.

Reply:
You should be able to use [errors.Is](https://golang.org/pkg/errors/#Is) to detect known error types.

Reply:
Trying to use it, but from my POV looks like unnecessary over complicating:
```go
	if s.emailAttributePath != "" {
		email, err := s.searchJSONForAttr(s.emailAttributePath, data.rawJSON)
		if !(xerrors.Is(err, ErrNoAttribute) ||
			xerrors.Is(err, ErrEmptyUserInfoResponse) ||
			xerrors.Is(err, ErrAttributeNotFound)) {
			s.log.Error("Error searching email attribute", "err", err)
		}
		if err == nil && email != "" {
			return email
		}
	}
```
We don't stop login in any of that cases, so probably debug logging of any error is ok. Also, we could change `searchJSONForAttr()` function and not return expected errors, like `ErrNoAttribute` or `ErrEmptyUserInfoResponse` and just return `ErrAttributeNotFound` or just `""` without error.

Reply:
The idea with using error level logging on unexpected errors is that we want to catch them, because they should indicate something is wrong (for example programming errors).

Reply:
I agree, it feels weird to check for expected errors. Again, if we're only interested in unexpected errors you should always fail on an error. With that said, error should only be returned from searchJSONForAttr if unmarshal or jmespath search fails, I think.

On another note, I've already went through trying to return errors from methods in oauth flow in this pr https://github.com/grafana/grafana/pull/14683#discussion_r315798458 and I wasn't satisfied how that ended up looking.

Reply:
@marefr The alternative is of course to only return unexpected errors from s.searchJSONForAttr. In that case, in this section of the code (calling said method) we know that errors are unexpected and the need for checking the error type falls away. The question then becomes what to do with unexpected errors.

@marefr What should our strategy be here for dealing with unexpected errors? Bubble or log an error?

Reply:
Unexpected should probably bubble up. But we can log them for now maybe and handle the bigger change in a separate PR?

Reply:
@marefr Yeah, I think that's the conclusion we've reached. To just error log unexpected errors for now, and bubble in a future PR.